File: internal/models/recording.go
package models

import "time"

// RecordingStatus is the lifecycle state of a session recording.
type RecordingStatus string

const (
	RecordingPending   RecordingStatus = "pending"   // metadata created, not yet writing
	RecordingActive    RecordingStatus = "active"    // capturing
	RecordingFinalized RecordingStatus = "finalized" // complete and playable
	RecordingFailed    RecordingStatus = "failed"    // capture errored
	RecordingDiscarded RecordingStatus = "discarded" // blob removed (retention/abort)
)

// Recording is the control-plane metadata for one captured session. The bytes
// live in a blob store keyed by StorageKey; this row is the queryable index and
// never holds secret material.
type Recording struct {
	ID             string `gorm:"primaryKey"`
	UserID         string `gorm:"index"`
	Username       string
	ConnectionID   string `gorm:"index"`
	ConnectionName string
	Protocol       string `gorm:"index"`
	RouteID        string
	StreamID       string
	Class          string `gorm:"index"` // terminal | desktop
	Format         string // asciicast_v2 | webm_canvas | ...
	Authoritative  bool
	Status         RecordingStatus `gorm:"index"`
	Title          string
	StartedAt      time.Time
	EndedAt        *time.Time
	DurationMS     int64
	Size           int64
	Checksum       string // sha256 hex of the finalized blob
	StorageKey     string
	Error          string
	ExpiresAt      *time.Time `gorm:"index"` // nil = retained indefinitely
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

func (Recording) TableName() string { return "recordings" }

File: web/src/types/projection/recording.ts
export const RecordingClass = {
  Terminal: "terminal",
  Desktop: "desktop",
} as const;
export type RecordingClass =
  (typeof RecordingClass)[keyof typeof RecordingClass];

export const RecordingFormat = {
  AsciicastV2: "asciicast_v2",
  WebmCanvas: "webm_canvas",
} as const;
export type RecordingFormat =
  (typeof RecordingFormat)[keyof typeof RecordingFormat];

export const RecordingPolicy = {
  Disabled: "disabled",
  Manual: "manual",
  Auto: "auto",
} as const;
export type RecordingPolicy =
  (typeof RecordingPolicy)[keyof typeof RecordingPolicy];

export interface RecordingCapability {
  class: RecordingClass;
  formats: RecordingFormat[];
  authoritative: boolean;
  inputCapture: boolean;
}

export const RecordingStatus = {
  Pending: "pending",
  Active: "active",
  Finalized: "finalized",
  Failed: "failed",
  Discarded: "discarded",
} as const;
export type RecordingStatus =
  (typeof RecordingStatus)[keyof typeof RecordingStatus];

export interface RecordingSummary {
  id: string;
  userId: string;
  username?: string;
  connectionId: string;
  connectionName?: string;
  protocol: string;
  class: RecordingClass;
  format: RecordingFormat;
  authoritative: boolean;
  status: RecordingStatus;
  title?: string;
  startedAt: string;
  endedAt?: string;
  durationMs: number;
  size: number;
}

export interface RecordingFilters {
  user?: string;
  connection?: string;
  protocol?: string;
  class?: RecordingClass;
  status?: RecordingStatus;
}
